text/0517-io-os-reform.md

```diff
@@ -447,7 +458,173 @@ counts, arguments to `main`, and so on).
+  instead eventually introducing a much richer
+  parsing/formatting/(de)serialization framework that works seamlessly
+  with `Reader` and `Writer`.
+
+  Such a framework is out of scope for this RFC, but the
+  endian-sensitive functionality will be provided elsewhere
+  (likely out of tree).
+
+* **The error type**. The traits currently use `IoResult` in their
+  return types, which ties them to `IoError` in particular. Besides
+  being an unnecessary restriction, this type prevents `Reader` and
+  `Writer` (and various adapters built on top of them) from moving to
+  `libcore` -- `IoError` currently requires the `String` type.
+
+  With associated types, there is essentially no downside in making
+  the error type generic.
+
+With those general points out of the way, let's look at the details.
+
+### `Read`
+[Read]: #read
+
+The updated `Reader` trait (and its extension) is as follows:
+
+```rust
+trait Read {
+    type Err; // new associated error type
+
+    // unchanged except for error type
+    fn read(&mut self, buf: &mut [u8]) -> Result<uint, Err>;
+}
+
+// extension trait needed for object safety
+trait ReadExt: Read {
+    fn bytes(&mut self) -> Bytes<Self> { ... }
+    fn chars<'r>(&'r mut self) -> Chars<'r, Self, Err> { ... }
+
+    fn read_to_end(&mut self) -> Result<Vec<u8>, Err> { ... }
+    fn read_to_string(&self) -> Result<String, Err> { ... }
+
+    ... // more to come later in the RFC
+}
+impl<R: Read> ReadExt for R {}
+```
+
+Following the
+[trait naming conventions](https://github.com/rust-lang/rfcs/pull/344),
+the trait is renamed to `Read` reflecting the single method it
+provides.
+
+The `read` method should not involve internal looping (even over
+errors like `EINTR`). It is intended to be atomic.
+
+#### Removed methods
+
+The proposed `Read` trait is much slimmer than today's `Reader`. The vast
+majority of removed methods are parsing/deserialization, which were
+discussed above.
+
+The remaining methods (`read_exact`, `read_at_least`, `push`,
+`push_at_least`) were removed for various reasons:
+
+* `read_exact`, `read_at_least`: these are somewhat more obscure
+  conveniences that are not particularly robust due to lack of
+  atomicity.
+
+* `push`, `push_at_least`: these are special-cases for working with
+  `Vec`, which this RFC proposes to replace with a more general
+  mechanism described next.
+
+To provide some of this functionality in a more composition way,
+extend `Vec<T>` with an unsafe method:
+
+```rust
+unsafe fn with_extra(&mut self, n: uint) -> &mut [T];
+```
+
+This method is equivalent to calling `reserve(n)` and then providing a
+slice to the memory starting just after `len()` entries. Using this
+method, clients of `Read` can easily recover the `push` method.
+
+### `Write`
+[Write]: #write
+
+The `Writer` trait is cut down to even smaller size:
+
+```rust
+trait Write {
+    type Err;
+    fn write(&mut self, buf: &[u8]) -> Result<uint, Err>;
+    fn flush(&mut self) -> Result<(), Err> { ... }
```

Comment:
BufferedWriters and especially a GZIPWriter or other encryption writer may
certainly want to do extra things on a flush and even a close.  And async
network file systems only can signal certain errors on close(), otherwise
performance would be abysmal.  Even 'local' filesystems are getting
complicated with delayed allocations.  So hopefully we can at least
establish that special, important stuff can happen at close() time (and
close should probably imply a flush).  That said, I'm not sure how to even
check that stout got closed right, even in C.

- So you can ignore errors on flush/drop like is done today.
https://github.com/rust-lang/rust/blob/master/src/libstd/old_io/buffered.rs#L240
 , although there are still some head scratchers:
https://github.com/rust-lang/rust/blob/master/src/libstd/old_io/buffered.rs#L208

- You can panic all the time, if a flush or close fails inside drop, and
the user didn't previously do a consuming close(self) or non-consuming
close(&mut self).  Having a non-consuming close is another runtime branch
for BufferedWriter.write_u8(), and more code to write and an extra flag.
But you still have other attributes from the type like .path.  Obviously a
double panic will turn into an abort() - e.g.you first hit an array out of
bounds, and then IO failed due to drop(close(flush())), so abort.  It's
questionable in that case if you want to .flush() since the panic was
caused by an array out of bounds [programmer error].  Ouch.

- You can panic if you aren't already panicing for another reason.   The
second failure is ignored or perhaps it could be added to a list on the
current panic object.  (aka java's suppressed exceptions).  This could in
theory, be logged somewhere.

So the question is perhaps even deeper: should a drop() cause a flush,
ever?  What about if a panic is happening?  It reminds me of the
PoisonedMutex - you don't want to 'release it business as usual style'
during unwinding.  So likewise with files, you don't want to 'release'
(e.g. flush) data if you don't have to.

So after all that rambling here's a strawman.

- add consuming .close(&self) that returns an error, and highly recommend
people check that, and punt how to do that on stdout.  Are people saying
this is not possible for some technical reason, due to nested objects?  The
reason for the consuming close is it makes normal runtime methods faster.
- note prominently that drop will not panic (unless perhaps due to EBADF,
 which is like an internal assert on something highly dangerous)
- but also note prominently that drop will also NOT CALL CLOSE OR FLUSH of
any nested objects, it will just drop them.  That means that if you drop a
bufferedwriter without closing, or at least flushing it, you won't see your
data, so you'd probably catch that at dev time.

- Karl


On Thu, Jan 29, 2015 at 1:30 PM, mahkoh <notifications@github.com> wrote:

> In text/0517-io-os-reform.md
> <https://github.com/rust-lang/rfcs/pull/576#discussion_r23795344>:
>
> > +```
> > +
> > +This method is equivalent to calling `reserve(n)` and then providing a
> > +slice to the memory starting just after `len()` entries. Using this
> > +method, clients of `Read` can easily recover the `push` method.
> > +
> > +### `Write`
> > +[Write]: #write
> > +
> > +The `Writer` trait is cut down to even smaller size:
> > +
> > +```rust
> > +trait Write {
> > +    type Err;
> > +    fn write(&mut self, buf: &[u8]) -> Result<uint, Err>;
> > +    fn flush(&mut self) -> Result<(), Err> { ... }
>
> Close can just be a nop for almost all Writers. If someone calls close on
> a File simply replace the fd by -1. Then no runtime checks are necessary.
>
> Of course, having the destructor panic if a file doesn't close properly
> makes no sense.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/576/files#r23795344>.
>